plugins/org.eclipse.n4js.smith/src/org/eclipse/n4js/smith/SerializingDataCollector.java

```diff
@@ -0,0 +1,154 @@
+/**
+ * Copyright (c) 2020 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.smith;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.locks.ReentrantLock;
+
+import com.google.common.base.Supplier;
+
+/**
+ * A collector decorator that will ensure that measurements are not created concurrently.
+ */
+class SerializingDataCollector extends DataCollector {
```

Comment:
Cool idea to solve the issue with such a wrapper!

But isn't it possible that this can introduce new deadlocks if the code being measured obtains locks on other resources? Not that critical, probably, because this new locking will only be active during debugging, but it's something we have to keep in mind.

Reply:
Hmm interesting. I think this could happen, if we'd have re-entrant invocations on the DataCollector (which is discouraged, but possible with the API).
Thread A passes by the disabled collector, aquires some other lock L2.
Thread B enabled the collector, acquires the collectors lock L1 and tries subsequently to work with L2.
Thread A is now reentrant on the collector and waits for L1 to become available. 

The configuration of the collectors does not offer this kind of control flow from the internals of a request back to the outer request manager, but yes, we should keep an eye on this. I'd like to wait a concrete deadlock situation to happen before revisiting the implementation.